pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarCommandSenderImpl.java

```diff
@@ -191,7 +191,18 @@ public void sendActiveConsumerChange(long consumerId, boolean isActive) {
             return;
         }
         cnx.ctx().writeAndFlush(
-                Commands.newActiveConsumerChange(consumerId, isActive),
+                Commands.newActiveConsumerChange(consumerId, isActive, null),
+                cnx.ctx().voidPromise());
+    }
+
+    @Override
+    public void sendActiveConsumerChange(long consumerId, String keySharedProps) {
+        if (!Commands.peerSupportsActiveConsumerListener(cnx.getRemoteEndpointProtocolVersion())) {
+            // if the client is older than `v12`, we don't need to send consumer group changes.
+            return;
+        }
+        cnx.ctx().writeAndFlush(
+                Commands.newActiveConsumerChange(consumerId, true, keySharedProps),
```

Comment:
Why we should combine the active consumer notification with the key hash ranges change notification? It looks like in general, we should not send active consumer notification to the consumers with Shared and Key_Shared subscription.

Reply:
My suggestion is to use a new message for the key hash ranges change events.

Reply:
[In this Question ](https://github.com/apache/pulsar/issues/6555#issuecomment-903172278). I see you answer him "We have a consumer events listener https://github.com/apache/pulsar/blob/6704f12104219611164aa2bb5bbdfc929613f1bf/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerEventListener.java, users need to register a listener for getting the new events of the consumer ". So I think it will be implements by a event, not a message